### PR TITLE
network status tracking code, tests, and changes to build file for both.

### PR DESCRIPTION
**Pull request by makdharma**

### CODE REVIEW

File: src/core/lib/iomgr/ev_poll_and_epoll_posix.c [hunk 1]
```diff
@@ -547,10 +547,13 @@ static int set_ready_locked(grpc_exec_ctx *exec_ctx, grpc_fd *fd,
 
 static void fd_shutdown(grpc_exec_ctx *exec_ctx, grpc_fd *fd) {
   gpr_mu_lock(&fd->mu);
-  GPR_ASSERT(!fd->shutdown);
```
Comment (ctiller) [inline src/core/lib/iomgr/ev_poll_and_epoll_posix.c]: Like we talked about - would prefer this to be in tcp_endpoint directly to simplify possible fd implementations (there will be many).

File: src/core/lib/iomgr/network_status_tracker.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+typedef struct endpoint_ll_node {
+  grpc_endpoint *ep;
+  struct endpoint_ll_node *next;
+} endpoint_ll_node;
+
+static endpoint_ll_node *head = NULL;
+
+// TODO(makarandd): Install callback with OS to monitor network status.
+void initialize_network_status_monitor() {
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: grpc_ prefix

File: src/core/lib/iomgr/network_status_tracker.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+void destroy_network_status_monitor() {
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: grpc_ prefix

File: src/core/lib/iomgr/network_status_tracker.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  for (endpoint_ll_node *curr = head; curr != NULL; ) {
+    endpoint_ll_node *next = curr->next;
+    gpr_free(curr);
+    curr = next;
+  }
+}
+
+void network_status_register_endpoint(grpc_endpoint *ep) {
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: grpc_ prefix

File: src/core/lib/iomgr/network_status_tracker.c [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    head = (endpoint_ll_node *)gpr_malloc(sizeof(endpoint_ll_node));
+    head->ep = ep;
+    head->next = prev_head;
+  }
+}
+
+// Walk the linked-list from head and execute shutdown. It is possible that
+// other threads might be in the process of shutdown as well, but that has
+// no side effect.
+void network_status_shutdown_all_endpoints() {
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: grpc_ prefix

File: src/core/lib/iomgr/ev_poll_and_epoll_posix.c (hunk 1 above)
Comment (makdharma) [inline src/core/lib/iomgr/ev_poll_and_epoll_posix.c]: Done. The changes are now in tcp_posix.c and tcp_windows.c.

File: src/core/lib/iomgr/network_status_tracker.c (hunk 2 above)
Comment (makdharma) [inline src/core/lib/iomgr/network_status_tracker.c]: Done

File: src/core/lib/iomgr/network_status_tracker.c (hunk 3 above)
Comment (makdharma) [inline src/core/lib/iomgr/network_status_tracker.c]: Done

File: src/core/lib/iomgr/network_status_tracker.c (hunk 4 above)
Comment (makdharma) [inline src/core/lib/iomgr/network_status_tracker.c]: Done

File: src/core/lib/iomgr/network_status_tracker.c (hunk 5 above)
Comment (makdharma) [inline src/core/lib/iomgr/network_status_tracker.c]: Done

**Comment by makdharma** [comment]
Test this please

**Comment by makdharma** [comment]
@ctiller - I addressed your last set of comments. Anything more needs to be done?

File: src/core/lib/iomgr/network_status_tracker.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ */
+
+#ifndef GRPC_CORE_LIB_IOMGR_NETWORK_STATUS_TRACKER_H
+#define GRPC_CORE_LIB_IOMGR_NETWORK_STATUS_TRACKER_H
+#include "src/core/lib/iomgr/endpoint.h"
+
+void grpc_network_status_register_endpoint(grpc_endpoint *ep);
+void grpc_network_status_shutdown_all_endpoints();
+#endif /* GRPC_CORE_LIB_IOMGR_NETWORK_STATUS_TRACKER_H */
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.h]: nit: add a newline before the `#endif`

File: src/core/lib/iomgr/tcp_posix.c [hunk 7]
```diff
@@ -107,7 +112,16 @@ static void tcp_handle_write(grpc_exec_ctx *exec_ctx, void *arg /* grpc_tcp */,
 
 static void tcp_shutdown(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep) {
   grpc_tcp *tcp = (grpc_tcp *)ep;
-  grpc_fd_shutdown(exec_ctx, tcp->em_fd);
+  gpr_mu_lock(&tcp->mu);
```
Comment (ctiller) [inline src/core/lib/iomgr/tcp_posix.c]: What happens if tcp_write or tcp_read are called concurrently or after tcp_shutdown?
Ideally we'd want them to fail immediately, currently I suspect we'd see some kind of crash.

Comment (dklempner) [inline src/core/lib/iomgr/tcp_posix.c]: I believe it should just work. It's supposed to work -- if your notification observes the inserted edge on shutdown, you should get an error and follow the !success path of tcp_handle_read.

This is no different than if you directly call shutdown concurrently/after a read or write.

The one requirement is that you don't do it twice -- our implementation only injects a single extra shutdown edge, so the second one can get stuck if it doesn't observe a socket error.

Comment (makdharma) [inline src/core/lib/iomgr/tcp_posix.c]: My understanding was that tcp_shutdown was designed to be called asynchronously. This PR only adds one more place from where tcp_shutdown might is called. The behavior after shutdown is unchanged.

In any case, a quick reading of the code suggests that reading/writing on already shutdown FDs will generate errors that are handled appropriately.

**Comment by ctiller** [comment]
I think we should at least add something to endpoint_tests.c that proves this for all endpoints.

```c
grpc_endpoint_read(... foo ...);
assert(foo_not_called);
grpc_endpoint_shutdown(...);
assert(foo_called_with_failure);
grpc_endpoint_read(... bar ...);
assert(bar_called_with_failure);
```

**Comment by ctiller** [comment]
Since this has languished and I need the endpoint idempotency for other changes, I'm going to look at that piece today.

**Comment by makdharma** [comment]
@ctiller please LGTM now that your idempotent change is in.

**Comment by ctiller** [comment]
Branch has conflicts. You'll need to at least back out the partial idempotent endpoint shutdown made here.

**Pull request closed without merging**